src/java.base/share/classes/java/lang/StringFormat.java

```diff
@@ -0,0 +1,320 @@
+/*
+ * Copyright (c) 2024, Alibaba Group Holding Limited. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.lang;
+
+import jdk.internal.math.DoubleToDecimal;
+import jdk.internal.util.HexDigits;
+import java.math.BigInteger;
+import java.util.Arrays;
+import java.util.Formatter;
+import java.util.IllegalFormatConversionException;
+import java.util.function.LongToIntFunction;
+import java.util.function.ToIntFunction;
+
+/**
+ * Utility class for string format fastpath
+ */
+final class StringFormat {
+    private static final char DECIMAL_INTEGER           = 'd';
+    private static final char HEXADECIMAL_INTEGER       = 'x';
+    private static final char HEXADECIMAL_INTEGER_UPPER = 'X';
+    private static final char STRING                    = 's';
+
+    static String format(String format, Object... args) {
+        if (args != null) {
+            int off = format.indexOf('%');
```

Comment:
nit: instead of not null check, should this short circuit for null/empty args?

```suggestion
        if (args == null || args.length == 0) {
            // no formatting to be done
            return format;
        }

        int off = format.indexOf('%');
```

Reply:
j.u.Formatter supports args == null

Reply:
When `args == null` or `args.length == 0`, then the format string can still contain `%%` or `%n`, and those will be formatted to `"%"` and [`System​.lineSeparator()`] respectively.

[`System​.lineSeparator()`]: https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/lang/System.html#lineSeparator()

Reply:
Yeah, if args is null or empty and there are format specifiers in the format string it should throw `IllegalFormatException`. To @liach ’s earlier comments, `String.format` likely needs more exhaustive unit tests covering various positive and negative test cases.